Detect overlapping detections

sgiehl opened the pull request:
### Description:

It seems like we have introduced some overlapping detections recently.
There are various detections that work correctly, when they are parsed by a specific client parser, but when parsed with all parsers another parser overwrites the results.
I've added some tests in this PR that use the fixtures used for specific parsers to parse them fully and check if the result is still the same.
There are currently 31 failures, where that doesn't match anymore.

I think we should try to get rid of overlapping detections. A user agent should either be detected as a browser or a mobile app. But not both. 

One of the problematic detections seems to be this one:
https://github.com/matomo-org/device-detector/blob/c49f7e93b80530e010b966558020663f788d5e20/regexes/client/mobile_apps.yml#L2480-L2483

We might need to rethink the order in which the parsers are added, so such generic matches are done more later.
On the other side could every parser contain generic detections, which might overwrite other results...

Or would it make sense to somehow mark generic regexes so they are only used if nothing else matches?

@liviuconcioiu @sanchezzzhak  Any thoughts / ideas on that ?

refs #7668

### Review

* [ ] [Functional review done](https://developer.matomo.org/guides/pull-request-reviews#functional-review-done)
* [ ] [Potential edge cases thought about](https://developer.matomo.org/guides/pull-request-reviews#potential-edge-cases-thought-about) (behavior of the code with strange input, with strange internal state or possible interactions with other Matomo subsystems)
* [ ] [Usability review done](https://developer.matomo.org/guides/pull-request-reviews#usability-review-done) (is anything maybe unclear or think about anything that would cause people to reach out to support)
* [ ] [Security review done](https://developer.matomo.org/guides/security-in-piwik#checklist)
* [ ] [Wording review done](https://developer.matomo.org/guides/pull-request-reviews#translations-wording-review-done)
* [ ] [Code review done](https://developer.matomo.org/guides/pull-request-reviews#code-review-done)
* [ ] [Tests were added if useful/possible](https://developer.matomo.org/guides/pull-request-reviews#tests-were-added-if-usefulpossible)
* [ ] [Reviewed for breaking changes](https://developer.matomo.org/guides/pull-request-reviews#reviewed-for-breaking-changes)
* [ ] [Developer changelog updated if needed](https://developer.matomo.org/guides/pull-request-reviews#developer-changelog-updated-if-needed)
* [ ] [Documentation added if needed](https://developer.matomo.org/guides/pull-request-reviews#documentation-added-if-needed)
* [ ] Existing documentation updated if needed

sgiehl committed:
Detect overlapping detections

sanchezzzhak commented:
I don't see any other way out but to add exceptions to the rule
fix for mobile-apps.yaml
```yaml
# Electron generic apps
- regex:  ' (?!(?:AppleWebKit|brave|Franz|Mailspring|Notion|Basecamp|Evernote|catalyst|ramboxpro|BlueMail|BeakerBrowser|TweakStyle|Colibri|Polypane|VibeMate|(?:d|LT|Glass|Sushi|Flash|OhHai)Browser|Sizzy))([a-z0-9]*)(?:-desktop|-electron-app)?/(\d+\.[\d.]+).*Electron/'
  name: '$1'
  version: '$2'
 ```
 fix for bos.yaml
 ```yaml
 # Generic detections
- regex: '[a-z0-9_-]*(?:(?<!cu|power[ _]|m[ _])bot(?![ _]TAB|[ _]?5[0-9]|[ _]Senior|[ _]Junior)|analyzer|appengine|archiver?|checker|collector|crawl|crawler|(?<!node-|uclient-|Mikrotik/\d\.[x\d] |electron-)fetch(?:er)?|indexer|inspector|monitor|(?<!Microsoft |banshee-)project(?!or)|(?<!Google Wap|Blue )proxy|research|resolver|robots|(?<!Cam)scanner|scraper|script|searcher|(?<!-)security|spider|study|transcoder|uptime|user[ _]?agent|validator)(?:[^a-z]|$)'
  name: 'Generic Bot'
```

sanchezzzhak commented on Tests/DeviceDetectorTest.php:
```diff
@@ -261,6 +261,89 @@ public function getFixtures(): array
+                $exception
+            );
+        }
+
```
it will be easier to find the problem

```php        
        $messageError = \sprintf("Fixture:\n%s\nResult:\n%s",
            \Spyc::YAMLDump($fixtureData, 2, 0),
            \Spyc::YAMLDump($uaInfo, 2, 0)
        );

       unset($fixtureData['headers']); // ignore headers in result
       unset($fixtureData['client']['family']);

        $this->assertArrayNotHasKey('bot', $uaInfo, $messageError);
        $this->assertEquals($fixtureData['client'], $uaInfo['client'], $messageError);
```

sanchezzzhak commented:
I haven't figured out how to fix it yet.
```
Fixture:
---
user_agent: Mozilla/5.0 (X11; Linux x86_64; rv:21.0) Gecko/20100101 SlimerJS/0.7
client:
  type: library
  name: SlimerJS
  version: "0.7"

Result:
---
user_agent: Mozilla/5.0 (X11; Linux x86_64; rv:21.0) Gecko/20100101 SlimerJS/0.7
os:
  name: GNU/Linux
  version: ""
  platform: x64
client:
  type: browser
  name: Firefox
  version: ""
  engine: Gecko
  engine_version: "21.0"
device:
  type: desktop
  brand: ""
  model: ""
os_family: GNU/Linux
browser_family: Firefox
```

```
Fixture:
---
user_agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 10.0; Win64; x64; Anonymisiert durch AlMiSoft Browser-Maulkorb 39663422; Trident/7.0; .NET4.0C; .NET4.0E; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; Tablet PC 2.0; Browzar)
client:
  type: browser
  name: Browzar
  version: ""
  engine: Trident
  engine_version: "7.0"
  family: Internet Explorer

Result:
---
user_agent: Mozilla/4.0 (compatible; MSIE 7.0; Windows NT 10.0; Win64; x64; Anonymisiert durch AlMiSoft Browser-Maulkorb 39663422; Trident/7.0; .NET4.0C; .NET4.0E; .NET CLR 2.0.50727; .NET CLR 3.0.30729; .NET CLR 3.5.30729; Tablet PC 2.0; Browzar)
os:
  name: Windows
  version: "10"
  platform: x64
client:
  type: mobile app
  name: Browser-Anonymizer
  version: ""
device:
  type: desktop
  brand: ""
  model: ""
os_family: Windows
browser_family: Unknown
```

liviuconcioiu commented:
Now, I see why I had a hard time getting the Electron regex working :(.

@sanchezzzhak idea with the updated regexes is great.

sanchezzzhak commented:
@sgiehl can I make the changes suggested above in this branch?

sgiehl commented:
Feel free to make changes to that branch. I currently haven't time to. 
Changing the regex should work. But we could also see if changing the priority here helps:
https://github.com/matomo-org/device-detector/blob/c2713baf2bb3052168ffa6d46159817387a7ed64/DeviceDetector.php#L198-L203

sanchezzzhak committed:
fix: added excludes browsers for electron regex
fix: added excludes apps and browsers for bot generic regex
feat(test): improved error output for DeviceDetectorTest::testParseClient method

sanchezzzhak committed:
fix: added exclude Browser-Maulkorb prefix for Browser-Anonymizer app
fix: added exclude SlimerJS prefix for Firefox browser
